service/matching/taskListManager.go

```diff
@@ -587,12 +591,30 @@ func (c *taskListManagerImpl) getIsolationGroupForTask(ctx context.Context, task
 		}
 		group, err := c.partitioner.GetIsolationGroupByDomainID(ctx, taskInfo.DomainID, partitionConfig, pollerIsolationGroups)
 		if err != nil {
+			// For a sticky tasklist, return StickyUnavailableError to let it be added to the non-sticky tasklist.
+			if err == partition.ErrNoIsolationGroupsAvailable && c.taskListKind == types.TaskListKindSticky {
+				return "", _stickyPollerUnavailableError
+			}
+			// if we're unable to get the isolation group, log the error and fallback to no isolation
 			c.logger.Error("Failed to get isolation group from partition library", tag.WorkflowID(taskInfo.WorkflowID), tag.WorkflowRunID(taskInfo.RunID), tag.TaskID(taskInfo.TaskID), tag.Error(err))
-		} else {
-			return group
+			return "", nil
+		}
+		// For a sticky tasklist, it is possible that when an isolation group is undrained, the tasks from one workflow is reassigned
```

Comment:
Why is this only on undrain? Can't it happen on drain as well?

Reply:
it can happen on drain as well. I didn't mention it because that's not surprising.